container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java

```diff
@@ -28,69 +28,76 @@
     private final ScheduledExecutorService executor =
             Executors.newScheduledThreadPool(1, ThreadFactoryFactory.getThreadFactory("deconstructor"));
 
-    private final int delay;
+    private final Duration delay;
 
     public Deconstructor(boolean delayDeconstruction) {
-        delay = delayDeconstruction ? 60 : 0;
+        this.delay = delayDeconstruction ? Duration.ofSeconds(60) : Duration.ZERO;
     }
 
 
     @Override
     public void deconstruct(Object component) {
+        // The mix of deconstructing some components on dedicated thread and some on caller thread could in theory violate
+        // ordering contraints between components.
         if (component instanceof AbstractComponent) {
             AbstractComponent abstractComponent = (AbstractComponent) component;
             if (abstractComponent.isDeconstructable()) {
-                executor.schedule(new DestructComponentTask(abstractComponent), delay, TimeUnit.SECONDS);
+                executor.schedule(new DestructComponentTask(abstractComponent), delay.getSeconds(), TimeUnit.SECONDS);
             }
         } else if (component instanceof Provider) {
-            log.info("Starting deconstruction of " + component);
+            // TODO Providers should most likely be deconstructed similarily to AbstractComponent
+            log.info("Starting deconstruction of provider " + component);
             ((Provider)component).deconstruct();
-            log.info("Finished deconstructing " + component);
+            log.info("Finished deconstructing of provider " + component);
         } else if (component instanceof SharedResource) {
+            log.info("Releasing container reference to resource " + component);
             // No need to delay release, as jdisc does ref-counting
-            log.info("Starting deconstruction of " + component);
             ((SharedResource)component).release();
-            log.info("Finished deconstructing " + component);
         }
     }
 
     private static class DestructComponentTask implements Runnable {
 
+        private final Random random = new Random(System.currentTimeMillis());
```

Comment:
Here you assume that the task will not be created at the same time between containers, right? Why is that less likely than calling the random method at the same time between containers?

Reply:
It is extremely unlikely that the `DestructorComponentTask` instances will be created at the exact same timestamp (<1ms)  across two nodes.

Reply:
They did not, but still 8 out of 10 containers or something restarted at the same time evidently.

Reply:
That could be caused by either a `Provider` or `SharedResource` throwing an exception. No way that 8 of 10 nodes created a `Random` instance at the exact same milli/nanosecond.